File: web/src/api/components/schemas/Category.ts
import { Subcategory, SubcategoryIO } from '../../components/schemas/Subcategory';
import { Option } from 'fp-ts/lib/Option';
import { Int, string, boolean, array, type } from 'io-ts';
import { optionFromNullable } from 'io-ts-types/lib/optionFromNullable';

export type Category = {
	id: Option<Int>;
	slug: Option<string>;
	name: Option<string>;
	blurb: Option<string>;
	coverMediaId: Option<Int>;
	imageUrl: Option<string>;
	sort: Option<Int>;
	published: Option<boolean>;
	subcategoryCount: Option<Int>;
	videoCount: Option<Int>;
	subNames: Option<Array<string>>;
	subcategories: Option<Array<Subcategory>>;
};
export const CategoryIO = type(
	{
		id: optionFromNullable(Int),
		slug: optionFromNullable(string),
		name: optionFromNullable(string),
		blurb: optionFromNullable(string),
		coverMediaId: optionFromNullable(Int),
		imageUrl: optionFromNullable(string),
		sort: optionFromNullable(Int),
		published: optionFromNullable(boolean),
		subcategoryCount: optionFromNullable(Int),
		videoCount: optionFromNullable(Int),
		subNames: optionFromNullable(array(string)),
		subcategories: optionFromNullable(array(SubcategoryIO)),
	},
	'Category',
);

File: backend/src/main/java/net/tape/model/Category.java
package net.tape.model;

import lombok.Getter;
import lombok.Setter;

import java.io.Serializable;

/** Domain model — framework-free POJO. */
@Getter
@Setter
public class Category implements Serializable {
    private Long id;
    private String slug;
    private String name;
    private String blurb;
    private Long coverMediaId;
    private String imageUrl;
    private int sort;
    private boolean published;
    private int subcategoryCount;
    private int videoCount;
    private java.util.List<String> subNames;
    private java.util.List<Subcategory> subcategories;
}
